tests/test_fs.py

```diff
@@ -147,3 +149,27 @@ def test_initialize_from_fsspec(self):
         fs, _, _ = fsspec.get_fs_token_paths(f"hf://{self.repo_id}:/data/text_data.txt")
         self.assertEqual(fs.repo_id, self.repo_id)
         self.assertEqual(fs.repo_type, "model")
+
+
+@pytest.mark.parametrize(
+    "path,expected",
+    [
+        # Repo type "model" by default
+        ("gpt2", {"repo_id": "gpt2", "repo_type": "model"}),
+        ("hf://gpt2", {"repo_id": "gpt2", "repo_type": "model"}),
+        # Parse without protocol
+        ("datasets/username/my_dataset", {"repo_id": "username/my_dataset", "repo_type": "dataset"}),
+        # Parse with hf:// protocol
+        ("hf://datasets/username/my_dataset", {"repo_id": "username/my_dataset", "repo_type": "dataset"}),
+        # Parse with revision
+        (
+            "hf://datasets/username/my_dataset@0123456789",
+            {"repo_id": "username/my_dataset", "repo_type": "dataset", "revision": "0123456789"},
+        ),
+        # Parse canonical repos (no namespace)
+        ("gpt2", {"repo_id": "gpt2", "repo_type": "model"}),
+        ("hf://gpt2", {"repo_id": "gpt2", "repo_type": "model"}),
```

Comment:
small bonus:
```suggestion
        ("hf://gpt2", {"repo_id": "gpt2", "repo_type": "model"}),
        ("datasets/squad", {"repo_id": "squad", "repo_type": "dataset"}),
        ("hf://datasets/squad", {"repo_id": "squad", "repo_type": "dataset"}),
```

Reply:
IIRC `RepoUrl` cannot parse canonical URLs that are prefixed with `dataset(s)/`, but I agree it would nice if it could

Reply:
Then let's at least document it as a "known failing case" (and open an issue if you think it's relevant). I'll update that

Reply:
I think we'll need this yes

Reply:
I fixed the parsing of canonical URLs in https://github.com/huggingface/hffs/pull/12/commits/e50baa23458e8df6588c1c5d842924119cd2f1f3.
Also opened an issue in `huggingface_hub` (https://github.com/huggingface/huggingface_hub/issues/1336) that I will fix right away but will be release only in a few weeks. In the meantime I have hardcoded a fix in `hffs`

Reply:
Addressed in https://github.com/huggingface/huggingface_hub/pull/1337.